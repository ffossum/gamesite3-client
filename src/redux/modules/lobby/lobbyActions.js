/* @flow */
import type { GameDataState, GamesState } from '../games/gamesReducer';

export const CREATE_GAME_REQUEST = 'lobby/create game request';
export const GAME_CREATED = 'lobby/game created';
export const GAME_UPDATED = 'lobby/game updated';
export const ENTER_LOBBY = 'lobby/enter';
export const EXIT_LOBBY = 'lobby/exit';
export const REFRESH_LOBBY = 'lobby/refresh';

export type EnterLobbyAction = {
  type: typeof ENTER_LOBBY,
};
export function enterLobby(): EnterLobbyAction {
  return {
    type: ENTER_LOBBY,
  };
}

export type ExitLobbyAction = {
  type: typeof EXIT_LOBBY,
};
export function exitLobby(): ExitLobbyAction {
  return {
    type: EXIT_LOBBY,
  };
}

export type RefreshLobbyAction = {
  type: typeof REFRESH_LOBBY,
  payload: GamesState,
};
export function refreshLobby(lobby: GamesState): RefreshLobbyAction {
  return {
    type: REFRESH_LOBBY,
    payload: lobby,
  };
}

export type CreateGameRequestAction = {
  type: typeof CREATE_GAME_REQUEST,
  payload: {
    userId: string,
  },
};

export function createGameRequest(userId: string): CreateGameRequestAction {
  return {
    type: CREATE_GAME_REQUEST,
    payload: {
      userId,
    },
  };
}

export type GameCreatedAction = {
  type: typeof GAME_CREATED,
  payload: GameDataState,
};

export function gameCreated(gameData: GameDataState): GameCreatedAction {
  return {
    type: GAME_CREATED,
    payload: gameData,
  };
}

export type GameUpdatedAction = {
  type: typeof GAME_UPDATED,
  payload: $Shape<GameDataState>,
};
export function gameUpdated(
  partialGameData: $Shape<GameDataState>,
): GameUpdatedAction {
  return {
    type: GAME_UPDATED,
    payload: partialGameData,
  };
}

export type LobbyAction =
  | EnterLobbyAction
  | ExitLobbyAction
  | RefreshLobbyAction
  | GameUpdatedAction
  | CreateGameRequestAction
  | GameCreatedAction;
