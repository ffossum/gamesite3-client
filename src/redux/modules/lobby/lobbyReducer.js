/* @flow */
import type { Action } from '../../actions';
import { GAME_CREATED, REFRESH_LOBBY } from './lobbyActions';
import { keys, union } from 'ramda';

type GameId = string;
export type LobbyState = {
  games: GameId[],
};

const initialState = {
  games: [],
};

export default function lobbyReducer(
  state: LobbyState = initialState,
  action: Action,
) {
  switch (action.type) {
    case REFRESH_LOBBY: {
      const gameIds = keys(action.payload);
      return {
        ...state,
        games: gameIds,
      };
    }
    case GAME_CREATED: {
      const game = action.payload;

      return {
        ...state,
        games: union(state.games, [game.id]),
      };
    }
    default:
      return state;
  }
}
