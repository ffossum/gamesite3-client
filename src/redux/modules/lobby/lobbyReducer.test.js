/* @flow */
/* eslint-env jest */
import lobbyReducer from './lobbyReducer';
import { refreshLobby, gameCreated } from './lobbyActions';

describe('lobby reducer', () => {
  test('initial state', () => {
    const initialState = lobbyReducer(undefined, { type: '@@INIT' });
    expect(initialState).toMatchSnapshot();
  });

  test('refresh lobby action completely replaces game list', () => {
    const initialState = {
      games: ['game0'],
    };

    const action = refreshLobby({
      game1: {
        id: 'game1',
        createdTime: '2017-06-19T20:33:46.834Z',
        host: 'user1',
        players: ['user1', 'user2'],
      },
      game2: {
        id: 'game2',
        createdTime: '2017-06-19T20:34:44.520Z',
        host: 'user2',
        players: ['user2', 'user3'],
      },
    });

    const state = lobbyReducer(initialState, action);

    expect(state.games).toHaveLength(2);
    expect(state.games).toContainEqual('game1');
    expect(state.games).toContainEqual('game2');
  });

  test('create game action adds game id to list', () => {
    const initialState = {
      games: ['game0'],
    };

    const action = gameCreated({
      id: 'game1',
      createdTime: '2017-06-19T20:33:46.834Z',
      host: 'user1',
      players: ['user1', 'user2'],
    });

    const state = lobbyReducer(initialState, action);
    expect(state.games).toHaveLength(2);
    expect(state.games).toContainEqual('game0');
    expect(state.games).toContainEqual('game1');
  });
});
