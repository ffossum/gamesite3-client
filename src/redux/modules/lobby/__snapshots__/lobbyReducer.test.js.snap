// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lobby reducer initial state 1`] = `
Object {
  "games": Array [],
}
`;
